### ISSUE TITLE
Doesn't show errors in files in /tmp on OS X

### ISSUE THREAD

**Issue by AgentME** [AtomLinter/linter-flow#64]
/tmp/foo/foo.js

```
/* @flow */

var x = 9;
console.log(x.length);
```

Assuming Atom is opened in /tmp/foo, and `flow init` has been run in the directory. Running `flow` shows the error, but linter-flow doesn't show any issues in the file /tmp/foo/foo.js. It reports 1 issue in the project in the status bar. Clicking the "1 issue" button causes /private/tmp/foo/foo.js to be opened, which shows the same contents, with a red squiggly line correctly appearing under `length` on the last line. Atom seems to believe they're separate files but they seem to be the same thing.

In OS X, /tmp appears to be a symlink to private/tmp, but I can't reproduce this issue with regular symlinked directories that I create.

**Comment by Arcanemagus** [comment]
Have you tried simply putting that folder into the ignore pattern in `linter` itself? This isn't really part of the scope of a linter plugin...

**Comment by AgentME** [comment]
I'm not sure I understand. Linter-flow doesn't report the errors, but running flow from the command line sees the file fine and shows the errors in it.

**Comment by Arcanemagus** [comment]
Sorry, I completely missed the fact that it's _not_ doing that, and that you want it to, instead of the opposite of that.

When you run `flow status --json` what is the path reported back by `flow`? I'm guessing that it's reporting back `/private/tmp/foo/foo.js` even though you are in `/tmp/foo/`.

**Comment by AgentME** [comment]
Yeah, it is:

```
[chris /tmp/foo]$ flow
test.js:3
  3: const x = foo;
               ^^^ identifier `foo`. Could not resolve name


Found 1 error
[chris /tmp/foo]$ flow status --json
{"passed":false,"errors":[{"message":[{"descr":"identifier `foo`","level":"warning","path":"/private/tmp/foo/test.js","line":3,"endline":3,"start":11,"end":13},{"descr":"Could not resolve name","level":"warning","path":"","line":0,"endline":0,"start":1,"end":0}],"kind":"infer"}],"version":" Mar  2 2016 14:35:09"}
```

**Comment by Arcanemagus** [comment]
Ah, this is an upstream `flow` issue then.

As this is a project scoped linter we are relying on `flow` to return the proper path in its messages to identify what files in the project the messages belong to. Can you file an issue on [their repo](https://github.com/facebook/flow) about this referencing this?

**Comment by qm3ster** [comment]
This seems to still be the case.
Is there a way I can use the plugin in its current state?

**Comment by Arcanemagus** [comment]
@qm3ster If you read the previous comments in the issue you can see that this is an upstream issue in `flow` that we have no control over, or ability to fix. Please file an issue over there as it seems that @AgentME didn't get a chance to.

**Comment by qm3ster** [comment]
Oh, I thought he surely must have filed it, I just couldn't find it because they have a lot of open issues. A lot.
I also imagined there might have been a workaround found in the meantime.
Perhaps with this behavior being consistent for such a long time, this package should adapt to `/private/tmp/foo/foo.js` being returned?
Finally, I am not confident enough in my understanding of the problem to file the issue, I am afraid.

**Comment by Arcanemagus** [comment]
Unfortunately we can't really work around this here.

We ask `flow` to run a lint, and the results it returns can have errors for _any_ file in the project, not just the file that the lint was triggered on. This means we have to rely on `flow` to return back the proper file path in order to associate it back with the right file inside Atom. Since `flow` is returning a path entirely different from what it should be, we have no idea what file it is talking about and aren't able to associate the messages with the open file.

The simple fix for this is to not try to lint projects with `flow` that are located in a symlinked directory (which `/tmp` apparently is on OS X :confused:).

**Comment by AgentME** [comment]
I don't think flow is doing anything wrong. Flow is always reporting the absolute paths of files, resolving all symlinks. Running `realpath /tmp` returns "/private/tmp".

I think the issue might've been Atom's fault. Maybe Atom usually resolved the absolute path of files, but didn't resolve /tmp. However, I can't reproduce the issue with linter-flow and Atom 1.7.2 now, so maybe that got fixed.

**Comment by qm3ster** [comment]
After messing about more, I started getting flow spawn ENOENT errors every time a linting would trigger.
Reinstalling any part separately did not work, it stayed the same.
But rebooting, not running Atom at all, 
`apm remove linter-flow`
`npm remove -g flow`
rebooting again
`npm install flow`
`apm installl linter-flow`
and only then opening Atom has finally, for the first time, made it work.
Life is good again.
Consider this a confirmation of compatibility of current Ubuntu 15.10, Atom 1.7.2, linter-flow and npm flow.

**Comment by Arcanemagus** [comment]
@AgentME I would consider it a bug in `flow` if you ask it to lint `/some/path/to/foo.js` and it reports back messages about `/a/completely/different/path/to/bar.js`. But I guess that's a matter of opinion... with _every_ other linter falling on the side of "Messages should be reported on what it was actually called on, not the de-referenced location"...

@qm3ster That sounds like a PATH issue, not an issue with `flow` reporting absolute paths instead of the current path. Glad to hear that it is working for you though!

**Comment by qm3ster** [comment]
Ok, I was completely stupid.
For some reason it only works when atom is started from console. Other linters do not have this problem.
The launcher calls **Exec=**`/opt/atom/atom %F` but I tried changing it to `atom %F` and sh -c 'atom %F' and permutations of all 3 without the %F parameter, but in all cases it's spawn flow ENOENT.
Calling it from console, in any current directory, or directly with a full file path passed makes it work every time.
Should I file this as a separate issue? It seems pretty concrete.
It definitely has to do with this package specifically, because it doesn't affect other linters that call systemwide tools.
On the other hand it seems like it is also indicative of an issue in Atom, since the launcher icon from the default installation should provide the same environment as launching from console.

**Comment by Arcanemagus** [comment]
Actually @qm3ster Apple had the bright idea to launch applications from the GUI with a _completely_ empty environment... so unfortunately it does _not_ have the same environment as being launched from a terminal.

We have several workarounds in place to attempt to get a useable environment when launching programs, but it's not perfect which is why there is an Executable Path setting that allows you to override the automatic detection.

**Comment by qm3ster** [comment]
This seems to be the case, yes.
I noticed that if I put `/home/qm3ster/.nvm/versions/node/v5.7.0/bin/flow` in **Executable Path**, running from the launcher gives the following instead:
```
Error: /usr/bin/env: node: No such file or directory
    at ChildProcess.<anonymous> (/home/qm3ster/.atom/packages/linter-eslint/node_modules/atom-linter/node_modules/sb-exec/lib/index.js:36:20)
```
(Running from `atom` is fine)
Turns out the same thing would affect linter-eslint, I just forever had `/home/qm3ster/.nvm/versions/node/v5.7.0` set as **Global Node Installation Path**, and it ships its own eslint or uses project-local ones. (This is part of their normal setup for global eslint. Marked *(Optional)*.)

So, for your project, I strongly suggest creating this setting too. (It seems that if I could manually point out both `node` and `flow` everything would go over smoothly in the apparently broken setup I've usually used.)

I also ask that you help me file this issue with Atom, since you as the package dev know much better what it is that Atom fails to provide to your package when ran through the launcher.

**Comment by Arcanemagus** [comment]
v1.7.0 already introduced some workarounds within Atom, it's not ideal, but it helps.

If you want to help debug why this in particular is failing you can file an issue on [consistent-env](https://github.com/steelbrain/consistent-env) as that is (eventually) what is providing the environment that is used to execute this package.

`linter-eslint` operates entirely differently from this package, and what you are seeing there doesn't really translate to here.

**Comment by qm3ster** [comment]
I could essentially reproduce this problem on that package, so I thought you could use the same crutch until this is fixed proper.
I currently shoved **Exec=**`env PATH="$PATH+:/home/mihail/.nvm/versions/node/v5.7.0/bin" /opt/atom/atom %F` into my **Atom.desktop**.
The problem seems to be with the fact that my node is installed not under root user but using nvm, so it doesn't make it into /etc/environment:
```shell
qm3ster@qm3ster-Z77:~$ echo $PATH
/home/qm3ster/.nvm/versions/node/v5.7.0/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games
qm3ster@qm3ster-Z77:~$ cat /etc/environment 
PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games"
```
